contracts/token/AssetToken.sol

```diff
@@ -0,0 +1,201 @@
+   */
+  function approvedFor(uint256 _tokenId) public view returns (address) {
+    return approvals[_tokenId];
+  }
+
+  /**
+  * @dev Transfers the ownership of a given token ID to another address
+  * @param _to address to receive the ownership of the given token ID
+  * @param _tokenId uint256 ID of the token to be transferred
+  */
+  function transfer(address _to, uint256 _tokenId) public onlyOwner(_tokenId) {
+    require(_to != address(0));
+    cleanApprovalAndTransfer(msg.sender, _to, _tokenId);
+  }
+
+  /**
+  * @dev Approves another address to claim for the ownership of the given token ID
+  * @param _to address to be approved for the given token ID
+  * @param _tokenId uint256 ID of the token to be approved
+  */
+  function approve(address _to, uint256 _tokenId) public onlyOwner(_tokenId) {
+    address owner = ownerOf(_tokenId);
+
+    require(_to != owner);
+    if(approvedFor(_tokenId) == 0 && _to == 0) return;
+
+    approvals[_tokenId] = _to;
+    Approval(owner, _to, _tokenId);
+  }
+
+  /**
+  * @dev Claims the ownership of a given token ID
+  * @param _tokenId uint256 ID of the token being claimed by the msg.sender
+  */
+  function takeOwnership(uint256 _tokenId) public {
+    address owner = ownerOf(_tokenId); // will revert if token ID does not exist
+
+    require(owner != msg.sender);
+    require(approvedFor(_tokenId) == msg.sender);
+
+    cleanApprovalAndTransfer(owner, msg.sender, _tokenId);
+  }
+
+  /**
+  * @dev Internal function to clean current approval and transfer the ownership of a given token ID
+  * @param _from address which you want to send tokens from
+  * @param _to address which you want to transfer the token to
+  * @param _tokenId uint256 ID of the token to be transferred
+  */
+  function cleanApprovalAndTransfer(address _from, address _to, uint256 _tokenId) internal {
+    delete approvals[_tokenId];
+    removeToken(_from, _tokenId);
+    addToken(_to, _tokenId);
+    Approval(_from, 0, _tokenId);
+    Transfer(_from, _to, _tokenId);
+  }
+
+  /**
+  * @dev Private function to add a token ID to the list of a given address
+  * @param _to address representing the new owner of the given token ID
+  * @param _tokenId uint256 ID of the token to be added to the tokens list of the given address
+  */
+  function addToken(address _to, uint256 _tokenId) internal {
+    tokenOwner[_tokenId] = _to;
+    uint256 length = balanceOf(_to);
+    ownedTokens[_to].push(_tokenId);
+    ownedTokensIndex[_tokenId] = length;
+  }
+
+  /**
+  * @dev Private function to remove a token ID from the list of a given address
+  * @param _from address representing the previous owner of the given token ID
+  * @param _tokenId uint256 ID of the token to be removed from the tokens list of the given address
+  */
+  function removeToken(address _from, uint256 _tokenId) internal {
+    uint256 length = balanceOf(_from);
+    if(length <= 1) delete ownedTokens[_from][_tokenId];
```

Comment:
I'd go with:

  require(length > 0);
  if (length == 1) {
    delete ownedTokens[_from][_tokenId];
    delete ownedTokens[_from];
  }

Reply:
good point, actually there's no need to handle any specific scenario for single element arrays